main/src/ca/uwaterloo/flix/language/phase/TreeShaker.scala

```diff
@@ -31,16 +34,137 @@ import ca.uwaterloo.flix.util.Validation._
   * (b) Appears in a fact or a rule as a filter/transfer function.
   * (c) Appears in a function which itself is reachable.
   */
+
 object TreeShaker extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
 
   /**
     * Performs tree shaking on the given AST `root`.
     */
   def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
 
-    // TODO: Implement.
+    /**
+      * A queue of function definitions to be processed recursively.
+      *
+      * For example, if the queue contains the entry:
+      *
+      * -   f
+      *
+      * it means that the function definition f should be considered to determine new reachable functions.
+      */
+    val queue: mutable.Queue[Definition.Constant] = mutable.Queue.empty
+
+    /**
+      * A map used to collect reachable functions (symbols and the corresponding definitions).
+      */
+    val reachableFunctions: mutable.Map[Symbol.DefnSym, Definition.Constant] = mutable.Map.empty
+
+    /**
+      * Returns true iff the function definition `defn` is initially reachable.
+      *
+      * That is, returns true iff `defn` satisfies one of the following:
+      *
+      *   (a) Appears in the global namespaces, takes zero arguments, and is not marked as synthetic.
+      *
+      *   (b) Appears in a fact or a rule as a filter/transfer function.
+      */
+    def firstReachable(defn: Definition.Constant): Boolean = {
```

Comment:
Any function symbol which appears in a fact or rule should be considered reachable.